src/db/mysql/batch.rs

```diff
@@ -170,24 +172,69 @@ pub fn do_append(
     db: &MysqlDb,
     batch_id: i64,
     user_id: HawkIdentifier,
-    _collection_id: i32,
+    collection_id: i32,
     bsos: Vec<params::PostCollectionBso>,
 ) -> Result<()> {
-    let inserts: Vec<_> = bsos
-        .into_iter()
-        .map(|bso: params::PostCollectionBso| {
-            let payload_size = bso.payload.as_ref().map(|p| p.len() as i64);
-            (
+    fn exist_idx(collection_id: i32, batch_id: i64, bso_id: &str) -> String {
+        format!(
+            "{collection_id}::{batch_id}::{bso_id}",
+            collection_id = collection_id,
+            batch_id = batch_id,
+            bso_id = bso_id,
+        )
+    };
+
+    #[derive(Debug, QueryableByName)]
+    struct ExistsResult {
+        #[sql_type = "Integer"]
+        batch: i32,
+        #[sql_type = "Integer"]
+        id: i32,
+    };
+
+    let mut existing = HashSet::new();
+    let mut inserts: Vec<_> = Vec::new();
+
+    for item in sql_query("SELECT batch, id FROM batch_upload_items where userid=@userid;")
```

Comment:
The Python version constructs a fancy INSERT ON DUPLICATE UPDATE that does both inserts and updates in one go.

This version is probably good enough for all our needs, but we might wanna log an issue that there's room for improvement

Reply:
Yep. Part of my approach here was to mirror the way that Spanner works. I'll agree it's less efficient, but it's a bit easier to keep in mind.